crate/server/src/database/redis/permissions.rs

```diff
@@ -1,196 +1,131 @@
+use std::collections::HashMap;
+
 use super::Redis;
 use crate::{
     database::database_traits::PermissionsTrait,
     error::{result::FResult, server::FindexServerError},
 };
 use async_trait::async_trait;
-use cosmian_crypto_core::bytes_ser_de::Serializable;
 use cosmian_findex_structs::{Permission, Permissions, WORD_LENGTH};
-use redis::{
-    aio::{ConnectionLike, ConnectionManager},
-    cmd, pipe, AsyncCommands, Pipeline, RedisError, ToRedisArgs,
-};
-use std::future::Future;
-use tracing::{debug, instrument, trace};
+use redis::AsyncCommands;
+use tracing::{instrument, trace};
 use uuid::Uuid;
 
-async fn transaction_async<
-    C: ConnectionLike + Clone,
-    K: ToRedisArgs,
-    T: Send,
-    F: Fn(C, Pipeline) -> Fut,
-    Fut: Future<Output = Result<Option<T>, RedisError>> + Send,
->(
-    mut cnx: C,
-    guards: &[K],
-    transaction: F,
-) -> Result<T, RedisError> {
-    loop {
-        cmd("WATCH").arg(guards).exec_async(&mut cnx).await?;
-        match transaction(cnx.clone(), pipe().atomic().to_owned()).await? {
-            None => continue,
-            Some(response) => {
-                return Ok(response);
-            }
-        }
-    }
+fn parse_permission_string(perm_str: &str) -> FResult<Permission> {
+    perm_str
+        .parse::<u8>()
+        .map_err(|e| {
+            FindexServerError::Unauthorized(format!("Invalid permission string: {perm_str}. {e}"))
+        })
+        .and_then(|p| {
+            Permission::try_from(p).map_err(|e| {
+                FindexServerError::Unauthorized(format!("Invalid permission value: {p}. {e}"))
```

Comment:
Failure to parse a permission is not a failure to authorize a request: change the error kind.